### PR TITLE
Add support for a sample_weight argument to fast_numba_auc().

### PR DESCRIPTION
**Pull request by Agade09**

### CODE REVIEW

File: fastauc/fast_auc.py [hunk 1]
```diff
@@ -44,21 +44,24 @@ def trapezoid_area(x1, x2, y1, y2):
     return dx * y1 + dy * dx / 2.0
 
 @numba.njit
-def fast_numba_auc(y_true: np.array, y_prob: np.array):
+def fast_numba_auc(y_true: np.array, y_prob: np.array,sample_weight=np.array([])):
```
Comment (diditforlulz273) [inline fastauc/fast_auc.py]: plz. a space after comma

File: fastauc/fast_auc.py [hunk 2]
```diff
@@ -44,21 +44,24 @@ def trapezoid_area(x1, x2, y1, y2):
...
     y_true = (y_true == 1)
 
     desc_score_indices = np.argsort(y_prob, kind="mergesort")[::-1]
     y_score = y_prob[desc_score_indices]
     y_true = y_true[desc_score_indices]
+    if len(sample_weight)>0:
```
Comment (diditforlulz273) [inline fastauc/fast_auc.py]: spaces around >

File: fastauc/fast_auc.py [hunk 3]
```diff
@@ -44,21 +44,24 @@ def trapezoid_area(x1, x2, y1, y2):
...
 
     prev_fps = 0
     prev_tps = 0
     last_counted_fps = 0
     last_counted_tps = 0
     auc = 0.0
     for i in range(len(y_true)):
-        tps = prev_tps + y_true[i]
-        fps = prev_fps + (1 - y_true[i])
+        weight = (sample_weight[i] if len(sample_weight)>0 else 1.0)
```
Comment (diditforlulz273) [inline fastauc/fast_auc.py]: spaces around >

File: fastauc/fast_auc.py [hunk 4]
```diff
@@ -44,21 +44,24 @@ def trapezoid_area(x1, x2, y1, y2):
...
+        tps = prev_tps + y_true[i]*weight
```
Comment (diditforlulz273) [inline fastauc/fast_auc.py]: spaces around * operator

File: fastauc/fast_auc.py [hunk 5]
```diff
@@ -44,21 +44,24 @@ def trapezoid_area(x1, x2, y1, y2):
...
+        fps = prev_fps + (1 - y_true[i])*weight
```
Comment (diditforlulz273) [inline fastauc/fast_auc.py]: spaces around * operator

File: fastauc/fast_auc.py [hunk 6]
```diff
@@ -68,7 +71,7 @@ def fast_numba_auc(y_true: np.array, y_prob: np.array):
     return auc / (prev_tps*prev_fps)
 
 
-def fast_auc(y_true: np.array, y_prob: np.array) -> Union[float, str]:
+def fast_auc(y_true: np.array, y_prob: np.array, sample_weight=None) -> Union[float, str]:
```
Comment (diditforlulz273) [inline fastauc/fast_auc.py:83]: update existing docstring with the new parameter description

File: fastauc/fast_auc.py [hunk 7]
```diff
@@ -84,12 +87,18 @@ def fast_auc(y_true: np.array, y_prob: np.array) -> Union[float, str]:
...
+    if sample_weight is not None:
+        sample_weight = sample_weight[desc_score_indices]
 
     distinct_value_indices = np.where(np.diff(y_score))[0]
     threshold_idxs = np.r_[distinct_value_indices, y_true.size - 1]
 
-    tps = np.cumsum(y_true)[threshold_idxs]
-    fps = 1 + threshold_idxs - tps
+    if sample_weight is not None:
+        tps = np.cumsum(y_true*sample_weight)[threshold_idxs]
```
Comment (diditforlulz273) [inline fastauc/fast_auc.py]: spaces around * operator

**Review by diditforlulz273 — CHANGES_REQUESTED** [review]

**Comment by diditforlulz273** [comment]
Also, if you add sample weights to two functions, please add it to the C++ one to make changes and logic similar

File: fastauc/cpp_auc.cpp [hunk 8]
```diff
@@ -1,22 +1,28 @@
...
     const float* b,
+    const float* sample_weight,
     const size_t len, 
-    std::vector<std::pair<bool,float>> &zipped)
+    std::vector<tuple_type> &zipped)
 {
     for(size_t i=0; i<len; ++i)
     {
-        zipped.push_back(std::make_pair(a[i], b[i]));
+        if constexpr(std::is_same<tuple_type,std::tuple<bool,float,float>>::value){
```
Comment (diditforlulz273) [inline fastauc/cpp_auc.cpp]: plz. spaces after commas :)

File: fastauc/cpp_auc.cpp [hunk 9]
```diff
@@ -26,42 +32,50 @@ double trapezoid_area(double x1, double x2, double y1, double y2) {
   return dx * y1 + dy * dx / 2.0;
 }
 
-float auc_kernel(float* ts, bool* st, size_t len) {
+template<typename tuple_type> float auc_kernel(float* ts, bool* st, size_t len,float* sample_weight) {
```
Comment (diditforlulz273) [inline fastauc/cpp_auc.cpp]: space after comma too

File: fastauc/cpp_auc.cpp [hunk 10]
```diff
@@ -26,42 +32,50 @@ double trapezoid_area(double x1, double x2, double y1, double y2) {
...
+  double fps = 0;
+  double tps = 0;
   double last_counted_fps = 0;
   double last_counted_tps = 0;
   double auc = 0.0;
   for(size_t i=0; i<zipped.size(); ++i) {
-    const double tps = prev_tps + zipped[i].first;
-    const double fps = prev_fps + (1 - zipped[i].first);
-    if( (i == zipped.size() - 1) || (zipped[i+1].second != zipped[i].second) ) {
+    if constexpr(std::is_same<tuple_type,std::tuple<bool,float,float>>::value){
```
Comment (diditforlulz273) [inline fastauc/cpp_auc.cpp]: spaces after commas plz.

File: fastauc/cpp_auc.cpp [hunk 11]
```diff
@@ -26,42 +32,50 @@ double trapezoid_area(double x1, double x2, double y1, double y2) {
...
   }
-  return auc / (prev_tps * prev_fps);
+  return auc / (tps * fps);
 }
 
 extern "C" {
-    float cpp_auc_ext(float* ts, bool* st, size_t len) {
-        return auc_kernel(ts, st, len);
+    float cpp_auc_ext(float* ts, bool* st, size_t len, float* sample_weight,size_t n_sample_weights) {
+        if(n_sample_weights>0){
```
Comment (diditforlulz273) [inline fastauc/cpp_auc.cpp]: spaces around > operator

File: fastauc/cpp_auc.cpp [hunk 12]
```diff
@@ -26,42 +32,50 @@ double trapezoid_area(double x1, double x2, double y1, double y2) {
...
+            return auc_kernel<std::tuple<bool,float,float>>(ts, st, len, sample_weight);
```
Comment (diditforlulz273) [inline fastauc/cpp_auc.cpp]: spaces after commas

File: fastauc/cpp_auc.cpp [hunk 13]
```diff
@@ -26,42 +32,50 @@ double trapezoid_area(double x1, double x2, double y1, double y2) {
...
+        }
+        else{
+            return auc_kernel<std::tuple<bool,float>>(ts, st, len, sample_weight);
```
Comment (diditforlulz273) [inline fastauc/cpp_auc.cpp]: spaces after commas

File: fastauc/fast_auc.py [hunk 14]
```diff
@@ -15,22 +15,25 @@ def __init__(self):
...
         self._handle.cpp_auc_ext.argtypes = [ndpointer(ctypes.c_float, flags="C_CONTIGUOUS"),
                                              ndpointer(ctypes.c_bool, flags="C_CONTIGUOUS"),
-                                             ctypes.c_size_t
+                                             ctypes.c_size_t,
+                                             ndpointer(ctypes.c_float, flags="C_CONTIGUOUS")
                                              ]
         self._handle.cpp_auc_ext.restype = ctypes.c_float
 
-    def roc_auc_score(self, y_true: np.array, y_score: np.array) -> float:
+    def roc_auc_score(self, y_true: np.array, y_score: np.array, sample_weight=np.array([],dtype=np.float32)) -> float:
```
Comment (diditforlulz273) [inline fastauc/fast_auc.py]: that's pretty bad actually - never declare mutable objects(like lists arrays etc) as default parameters in python:

Python’s default arguments are evaluated once when the function is defined, not each time the function is called (like it is in say, Ruby). This means that if you use a mutable default argument and mutate it, you will and have mutated that object for all future calls to the function as well.

You can find code examples here:
[click](https://www.30secondsofcode.org/articles/s/python-mutable-default-arguments)

File: fastauc/fast_auc.py [hunk 15]
```diff
@@ -71,12 +74,13 @@ def fast_numba_auc(y_true: np.array, y_prob: np.array,sample_weight=np.array([])
     return auc / (prev_tps*prev_fps)
 
 
-def fast_auc(y_true: np.array, y_prob: np.array, sample_weight=None) -> Union[float, str]:
+def fast_auc(y_true: np.array, y_prob: np.array, sample_weight: np.array=np.array([])) -> Union[float, str]:
```
Comment (diditforlulz273) [inline fastauc/fast_auc.py]: that's pretty bad actually - never declare mutable objects(like lists arrays etc) as default parameters in python:

Python’s default arguments are evaluated once when the function is defined, not each time the function is called (like it is in say, Ruby). This means that if you use a mutable default argument and mutate it, you will and have mutated that object for all future calls to the function as well.

You can find code examples here:
[click](https://www.30secondsofcode.org/articles/s/python-mutable-default-arguments)

File: fastauc/fast_auc.py [hunk 16]
```diff
@@ -44,13 +47,13 @@ def trapezoid_area(x1, x2, y1, y2):
     return dx * y1 + dy * dx / 2.0
 
 @numba.njit
-def fast_numba_auc(y_true: np.array, y_prob: np.array,sample_weight=np.array([])):
+def fast_numba_auc(y_true: np.array, y_prob: np.array, sample_weight: np.array=np.array([])):
```
Comment (diditforlulz273) [inline fastauc/fast_auc.py]: that's pretty bad actually - never declare mutable objects(like lists arrays etc) as default parameters in python:

Python’s default arguments are evaluated once when the function is defined, not each time the function is called (like it is in say, Ruby). This means that if you use a mutable default argument and mutate it, you will and have mutated that object for all future calls to the function as well.

You can find code examples here:
[click](https://www.30secondsofcode.org/articles/s/python-mutable-default-arguments)

**Review by diditforlulz273 — CHANGES_REQUESTED** [review]
I ask for code style changes and, more importantly, to not use mutable default arguments - the previous approach with None is the classic actually.

**Comment by Agade09** [comment]
I struggled to get numba to take None. That is the reason I gave a numpy array as default. I pushed a workaround by having a wrapper python function before calling numba. I didn't find better.

**Comment by diditforlulz273** [comment]
@Agade09 I think it would be faster if I correct some stuff myself, please re-PR the current code to the new branch I created, agade_2

**Pull request closed without merging**